File: macos/PackMan/Sources/PackMan/Models/AppState.swift
import Foundation

enum SourceScanState: Sendable, Equatable {
    case disabled
    case notScanned
    case waiting
    case probing(startedAt: Date)
    case scanning(phase: SourcePhase, startedAt: Date)
    case succeeded(updateCount: Int, completedAt: Date)
    case partial(updateCount: Int, issues: [SourceIssue], completedAt: Date)
    case unavailable(SourceIssue, completedAt: Date)
    case failed(SourceIssue, completedAt: Date)
    case cancelled(completedAt: Date)

    var hasIssue: Bool {
        switch self {
        case .partial, .unavailable, .failed, .cancelled: return true
        default: return false
        }
    }

    var isUnavailable: Bool {
        if case .unavailable = self { return true }
        return false
    }

    var issues: [SourceIssue] {
        switch self {
        case .partial(_, let issues, _): return issues
        case .unavailable(let issue, _), .failed(let issue, _): return [issue]
        default: return []
        }
    }
}

enum AppOperation: Sendable, Equatable {
    case idle
    case scanning(completed: Int, total: Int)
    case updating(current: String?, completed: Int, total: Int)
    case cancelling

    var isBusy: Bool { self != .idle }
}

enum ScanSummary: Sendable, Equatable {
    case notStarted
    case running
    case updatesAvailable(Int)
    case updatesCompleted(Date)
    case upToDate(Date)
    case completedWithIssues(updateCount: Int, issueCount: Int, completedAt: Date)
    case allUnavailable(Date)
    case noSources
    case cancelled(Date)
}

struct UpdateRunSummary: Sendable, Equatable {
    var updated = 0
    var failed = 0
    var cancelled = 0
    var verificationFailed = 0

    var total: Int { updated + failed + cancelled + verificationFailed }
}

struct LogEntry: Identifiable, Sendable, Equatable {
    enum Level: Sendable {
        case info
        case output
        case warning
        case error
        case success
    }

    let id: Int
    let timestamp: Date
    let level: Level
    let scope: String?
    let stream: ProcessOutputStream?
    let message: String
}

File: windows/src/PackMan/Models/AppState.cs
using CommunityToolkit.Mvvm.ComponentModel;
using PackMan.Services;

namespace PackMan.Models;

public enum SourceScanStatus
{
    Disabled,
    NotScanned,
    Waiting,
    Probing,
    Scanning,
    Installing,
    Succeeded,
    Partial,
    Unavailable,
    Failed,
    Cancelled,
}

public partial class SourceState : ObservableObject
{
    [ObservableProperty] private SourceScanStatus _status = SourceScanStatus.NotScanned;
    [ObservableProperty] private SourcePhase _phase = SourcePhase.Probing;
    [ObservableProperty] private DateTimeOffset? _startedAt;
    [ObservableProperty] private DateTimeOffset? _completedAt;
    [ObservableProperty] private int _updateCount;
    [ObservableProperty] private IReadOnlyList<SourceIssue> _issues = [];

    public bool HasIssue => Status is SourceScanStatus.Partial or SourceScanStatus.Unavailable
        or SourceScanStatus.Failed or SourceScanStatus.Cancelled;

    public void Set(SourceScanStatus status, SourcePhase phase = SourcePhase.Probing,
        int updateCount = 0, IReadOnlyList<SourceIssue>? issues = null)
    {
        Status = status;
        Phase = phase;
        UpdateCount = updateCount;
        Issues = issues ?? [];
        if (status is SourceScanStatus.Probing or SourceScanStatus.Scanning or SourceScanStatus.Installing)
            StartedAt ??= DateTimeOffset.Now;
        if (status is SourceScanStatus.Succeeded or SourceScanStatus.Partial or SourceScanStatus.Unavailable
            or SourceScanStatus.Failed or SourceScanStatus.Cancelled)
            CompletedAt = DateTimeOffset.Now;
        OnPropertyChanged(nameof(HasIssue));
    }
}

public enum AppOperationKind { Idle, Scanning, Updating, Installing, CleaningCache, Cancelling, UpdatingApp }
public enum ScanSummaryKind
{
    NotStarted,
    Running,
    UpdatesAvailable,
    UpdatesCompleted,
    UpToDate,
    CompletedWithIssues,
    AllUnavailable,
    NoSources,
    Cancelled,
}

public sealed record UpdateRunSummary(int Updated, int Failed, int Cancelled, int VerificationFailed);

public enum LogLevel { Info, Output, Warning, Error, Success }
public sealed record LogEntry(
    long Id,
    DateTimeOffset Timestamp,
    LogLevel Level,
    string Message,
    string? Scope = null,
    ProcessOutputStream? Stream = null)
{
    public string DisplayText => $"[{Timestamp:HH:mm:ss}]"
        + (Scope is null ? "" : $" [{Scope}]")
        + (Stream == ProcessOutputStream.StandardError ? " [stderr]" : "")
        + $" {Message}";
}
